zreader.go

```diff
@@ -32,6 +35,10 @@ const (
 	zGzip
 	zZip
 	zZstd
+	zZlib
+	zLz4Frame
+	zLz4Block // TODO: Can't use a magic byte approach for this
```

Comment:
I'm debating in my mind whether to (a) fall back on just trying `foo.NewReader` and trying something else if the error is not nil; or (b) *always* trying that strategy, instead of checking the magic numbers. (I'm assuming that for each `foo`, `foo.NewReader` already includes a magic number or other relevant check, as well as anything else relevant.)

But (b) raises the thorny issue of polyglot files, what order to try the modules in, multiple disk seeks (gotta seek back to the beginning for each attempt), et c. So, I dunno. What do you think?

Reply:
Q: By polyglot do you mean multiple compressed formats in a single file, or somethign ese?

What I was thinking was maybe something like a cascading search:
1. If we can easily identify from magic bytes, use that
2. Ask for the first full "block" of data (e.g. 1-4k bytes). Then use whatever sequence of `NewReader`s for our non-magic modes on our `Peek`'d bytes, and and try decoding using our set of "non-identifable" modes. E.g. `zLZ4Block` or `flate`. I'm assuming they'll return useful errors if they run on incorrectly composed data if we try and get like 1 byte of uncompressed data out of each.
    * E.g.`zTypeFromBytes` could return `ErrNeedMoreData`, and our logic is then to either run it again with more data or call something `ZtypeFromBytesForNonMagicSources` or w/e that has the `tryLZ4Block or tryFlake or [...]`

I think the way this might get into trouble is if there's any compression formats who's structure is similar enough to be readable by multiple different readers, but I don't know there layout well enough to know how likely that is xD

I do think where we can, it makes sense to rely on strong signals of what a stream is.

Reply:
Polyglot: Files that can be parsed as more than 1 format, e.g. https://arstechnica.com/information-technology/2008/08/newly-found-hybrid-attack-embeds-java-applet-in-gif-file/. I do kind of think/hope that with the formats we're talking about right now, the likelihood of that happening is low. So yeah I think your plan makes sense.

Reply:
Hmm, the library could check and fail (or give the clients a config option) if multiple readers are able to handle the data.